akka-stream/src/main/scala/akka/stream/impl/CompletedPublishers.scala

```diff
@@ -65,7 +69,51 @@ private[akka] final case class SingleElementPublisher[T](value: T, name: String)
     } catch {
       case _: SpecViolation ⇒ // nothing we can do
     }
-  def apply[T]: Publisher[T] = this.asInstanceOf[Publisher[T]]
+
+  override def toString: String = name
+}
+
+/**
+ * INTERNAL API
+ */
+private[akka] final case class TriggeredOptionPublisher[T](
+  promise: Promise[Option[T]],
+  name: String)(implicit ec: ExecutionContext) extends Publisher[T] {
+  import ReactiveStreamsCompliance._
+
+  private[this] class TriggeredOptionSubscription(subscriber: Subscriber[_ >: T]) extends Subscription {
+    private[this] var done: Boolean = false
+    override def cancel(): Unit = {
+      done = true
+      promise.trySuccess(None)
+    }
+
+    override def request(elements: Long): Unit = {
+      if (elements < 1) rejectDueToNonPositiveDemand(subscriber)
+      if (!done) {
+        done = true
+        promise.future foreach {
+          case Some(v) ⇒
+            tryOnNext(subscriber, v)
+            tryOnComplete(subscriber)
+          case None ⇒
+            tryOnComplete(subscriber)
+        }
```

Comment:
If you use a Sink.head, which is also actorless, then I think it will be:
 - materializer calls subscribe
 - subscribe calls onSubscribe on "head"
 - head requests and assume it has a bug (requests -1 for example)
 - exception is thrown. We are still on the materializing thread, which throws it back to caller. 

Am I missing something here?